### PR TITLE
134 time varying covariates

### PR DESCRIPTION
**Pull request by melodiemonod**
- [x] Extended cox model (with time varying hazard assuming time varying covariates not time varying parameters)
- [x] Survival model (approximation of the continuous log likelihood, allowing  time varying covariates and time varying parameters)
- [x] Add unittest for each new fcts

### CODE REVIEW

**Comment by melodiemonod** [comment]
Branch is ready for review

File: src/torchsurv/loss/cox.py [hunk 1]
```diff
@@ -536,18 +709,21 @@ def survival_function(
...
-
-        # Compute individual survival functions
-        individual_survival_strata = bs_strata.unsqueeze(0) ** torch.exp(new_log_hz_strata).unsqueeze(1)
+        time_stratum = baseline_survival_strata["time"]
+        bs_stratum = baseline_survival_strata["baseline_survival"]
 
         # Index of the largest element in time that is ≤ new_time
-        time_index = torch.searchsorted(time_strata, new_time, right=True) - 1
+        time_index = torch.searchsorted(time_stratum, new_time, right=True) - torch.tensor(1)
         time_index[time_index == -1] = 0
```
Comment (tcoroller) [inline src/torchsurv/loss/cox.py]: @melodiemonod  c'est pas ouf cette ligne 😆 

What is the `-1`, last item? Do you think you can either comment or reformulate in more robust manner?
Also maybe define the 0 as tensor(0) to prevent type mixing

File: docs/notebooks/metrics.md [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+$$
+\begin{align*}
+\text{AUC}^{\mathbb{I}/\mathbb{D}}(t) &= p(q_i(t)>q_j(t)|dN_i(t) = 1, N_j(t) = 0) \\
+& = p(q_i(t)>q_j(t)|X_i = t, X_j >t).
+\end{align*}
+$$
+
+It represents the probability that the model correctly identifies which of two comparable samples experiences an event at time $t$ based on their predicted risk scores.
+
+The proof of the probabilitic interpretation of $AUC^{\mathbb{I}/\mathbb{D}}$ is similar to that in the binary response context.
```
Comment (SoniaDem) [inline docs/notebooks/metrics.md]: typo: probabilistic

File: docs/notebooks/metrics.md [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+There exist three main limits of the AUC and C-index (Hartman et al., 2023):
+
+* **Ties in predicted risk score (C-index)**:
+A model that assigns the same risk score to individuals with extremely different survival experiences is not performing well, but the C-Index does not detect this inadequacy.
+On the other hand, a model that assigns the same risk score to individuals with very similar survival experiences may be appropriately capturing the similarities in underlying risk.
+While these two scenarios have very different interpretations, they are treated as the same in the C-Index calculation, since each of these pairs contributes a score of 0.5.
+
+* **Time dependence (AUC and C-index)**:
+Time-to-event outcome is dichotomized at each time point, which potentially generates many comparable pairs that are difficult to discriminate and are not clinically meaningful.
+For example, if a individual experiences the event of interest on day t of the study, and another individual, with similar underlying risk, experiences the event shortly afterwards on day t + 1, then these two individuals would be deemed comparable according to the time-dependent C-Index definition.
```
Comment (SoniaDem) [inline docs/notebooks/metrics.md]: a -> an individual

File: docs/notebooks/metrics.md [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+Below, we summarize the pros and cons of the C-index, AUC incident/dynamic, and AUC cumulative/dynamic (Lambert and Chevret, 2016).
+
+|  | C-index | AUC cumulative/dynamic| AUC incident/dynamic|
+|----------|----------|----------|-----------|
+| Pros | Global measure that does not depend on time | More clinically relevant, where one wishes to predict individuals who will have failed by time $t$. <br> Proper for time-dependent prediction. | Integral has probabilistic interpretation of the C-index. <br> Natural companion for instantaneous hazard models. <br>Proper for time-dependent prediction. |
+| Cons | Not proper for time-dependent prediction (e.g., 10-years mortality). | Integral does not have a nice probabilistic interpretation. <br> Requires a cumulative risk score (e.g., 1-survival function or cumulative hazard). | Require the exact event times to be recorded. <br> All subjects that failed before $t$ are ignored.|
```
Comment (SoniaDem) [inline docs/notebooks/metrics.md]: require -> requires (the exact event times...)

File: docs/notebooks/regression_time_varying.ipynb [hunk 5]
```diff
@@ -0,0 +1,1268 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "a07aa9fc",
+   "metadata": {},
+   "source": [
+    "# Survival Regression with Time-Varying Covariates\n",
+    "\n",
+    "In this notebook, we use `TorchSurv` to train a model that predicts relative risk given time-varying covariates. We use a the `heart` data set from the `survival` package in `R`. After training the model, we evaluate the predictive performance using evaluation metrics implemented in `TorchSurv`.\n",
```
Comment (SoniaDem) [inline docs/notebooks/regression_time_varying.ipynb]: remove 'a' -> we use the heart dataset ...

File: docs/notebooks/regression_time_varying.ipynb [hunk 6]
```diff
@@ -0,0 +1,1268 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "fa54b235",
+   "metadata": {},
+   "source": [
+    "### 3. Cox Model\n",
+    "\n",
+    "In this section, we use the [Extended Cox proportional hazards model](../_autosummary/torchsurv.loss.cox.html). Given covariates $x_{i}(t)$, a vector of size $p$, the hazard of patient $i$ has the form\n",
+    "$$\n",
+    "\\lambda (t) =\\lambda_{0}(t)\\: \\theta_i(t)\n",
```
Comment (SoniaDem) [inline docs/notebooks/regression_time_varying.ipynb]: optional: lambda_i(t) on the left?

File: docs/notebooks/regression_time_varying.ipynb [hunk 7]
```diff
@@ -0,0 +1,1268 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "c015d74d",
+   "metadata": {},
+   "source": [
+    "#### 3.3. Evaluation metrics\n",
+    "\n",
+    "We evaluate the predictive performance of the model using \n",
+    "\n",
+    "* the [concordance index](../_autosummary/torchsurv.metrics.cindex.html) (C-index), which measures the the probability that a model correctly predicts which of two comparable samples will experience an event first based on their estimated risk scores,\n",
```
Comment (SoniaDem) [inline docs/notebooks/regression_time_varying.ipynb]: measures the ... (double the here)

File: docs/notebooks/metrics.md [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+$$
+\begin{align*}
+AUC^{\mathbb{C}/\mathbb{D}}(t) &= p(q_i(t)>q_j(t)|N_i(t) = 1, N_j(t) = 0) \\
+&= p(q_i(t)>q_j(t)|X_i \leq t, X_j >t).
+\end{align*}
+$$
+
+It represents the probability that the model correctly identifies which of two comparable samples experiences an event before or at time $t$ based on their predicted risk scores.
+
+The proof of the probabilitic interpretation of $AUC^{\mathbb{C}/\mathbb{D}}$ is similar to that in the binary response context.
```
[the remark this answers is not in the document]
Comment (melodiemonod) [reply docs/notebooks/metrics.md]: done

File: docs/notebooks/metrics.md [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+\end{align*}
+$$
+
+It represents the probability that the model correctly identifies which of two comparable samples experiences an event at time $t$ based on their predicted risk scores.
+
+The proof of the probabilitic interpretation of $AUC^{\mathbb{I}/\mathbb{D}}$ is similar to that in the binary response context.
+
+#### 3.3 Extension of the C-index for time-to-event response
+
+The C-index is the probability that a model correctly predict which of two comparable samples experiences an event first based on their predicted risk scores. It is defined by:
```
[the remark this answers is not in the document]
Comment (melodiemonod) [reply docs/notebooks/metrics.md]: done

File: docs/notebooks/metrics.md [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+
+$$
+C = p(q_i(X_i) > q_j(X_i) | X_i < X_j)
```
[the remark this answers is not in the document]
Comment (melodiemonod) [reply docs/notebooks/metrics.md]: Nop, It represents the probability that, at the time when $i$ experiences the event,
            the risk score of subject $i$ exceeds the risk score of subject $j$,
            given that subject $i$ experiences the event before subject $j$. 

I've added a explanation there and also in the metrics/cindex docstring and added the proper citation (Heagerty2005)

File: docs/notebooks/metrics.md [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+- Use **AUC incidence/dynamic** when your risk score measures incidence (e.g., an instantaneous hazard).
+- Use **AUC cumulative/dynamic** when your risk score measures prevalence (e.g., a cumulative hazard).
+
+As a rule of thumb:
+- **Incidence-based questions** → **AUC incidence/dynamic**
+- **Prevalence-based questions** → **AUC cumulative/dynamic**
+
+#### Should I use the AUC or the C-index?
+
+The AUC is time-dependent and provide a measure of predictive performance at a specific time point. In contrast, the C-index offers a global assessment of a fitted survival model over the entire observational period. It is recommended to use the AUC instead of the C-index for time-dependent predictions (e.g., 10-year mortality), as AUC is proper in this context, while the C-index is not (Blanche et al., 2018).
```
[the remark this answers is not in the document]
Comment (melodiemonod) [reply docs/notebooks/metrics.md]: done

File: docs/notebooks/loss.md [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+
+**Avoid if:** interpretability or analytical tractability is a priority, or your dataset is small.
+
+
+
+### Summary Table
+
+| Model | Hazard Form | Time-varying Risk | Closed-form Likelihood | Best Used When... |
+|--------|--------------|-------------------|------------------------|-------------------|
+| **Cox** | $\lambda_i(t) = \lambda_0(t)\exp(f_{\theta}(\mathbf{x}_i))$ | ✗ (proportional only) | ✓  | You expect proportional hazards |
```
[the remark this answers is not in the document]
Comment (melodiemonod) [reply docs/notebooks/loss.md]: thanks, that is fixed

File: docs/notebooks/helpers_time_varying.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    for idx, _id in enumerate(ids):
+        id_idx = id == _id
+
+        # Which row within this id to use
+        time_idx = torch.searchsorted(start[id_idx], time, right=True) - 1
+
+        # Slice out the correct hazard value
+        value = log_hz_short[id_idx][time_idx].squeeze()
+
+        # Repeat the value across the entire row (your original behavior)
```
[the remark this answers is not in the document]
Comment (melodiemonod) [reply docs/notebooks/helpers_time_varying.py]: I;ve clarified int he text, please @SoniaDem  can you let me know if it's clearer? If not let me know how you would do it maybe by showing your code practice (don't hesitate to create a new notebook / or just .py) this would be very helpful. thanks!

**Comment by tcoroller** [comment]
I will do it Monday. Back from leave this week; getting back to this PR asap
